src/Microsoft.AspNetCore.Server.IIS/Core/IISHttpContext.FeatureCollection.cs

```diff
@@ -290,6 +294,42 @@ async Task<Stream> IHttpUpgradeFeature.UpgradeAsync()
             return new DuplexStream(RequestBody, ResponseBody);
         }
 
+        Task<X509Certificate2> ITlsConnectionFeature.GetClientCertificateAsync(CancellationToken cancellationToken)
+        {
+            return Task.FromResult(((ITlsConnectionFeature)this).ClientCertificate);
+        }
+
+        unsafe X509Certificate2 ITlsConnectionFeature.ClientCertificate
+        {
+            get
+            {
+                if (_certificate == null &&
+                    NativeRequest->pSslInfo != null &&
+                    NativeRequest->pSslInfo->pClientCertInfo != null &&
+                    NativeRequest->pSslInfo->pClientCertInfo->pCertEncoded != null &&
+                    NativeRequest->pSslInfo->pClientCertInfo->CertEncodedSize != 0)
+                {
+                    try
+                    {
+                        // Based off of from https://referencesource.microsoft.com/#system/net/System/Net/HttpListenerRequest.cs,1037c8ec82879ba0,references
+                        var rawCertificateCopy = new byte[NativeRequest->pSslInfo->pClientCertInfo->CertEncodedSize];
+                        Marshal.Copy((IntPtr)NativeRequest->pSslInfo->pClientCertInfo->pCertEncoded, rawCertificateCopy, 0, rawCertificateCopy.Length);
+                        _certificate = new X509Certificate2(rawCertificateCopy);
+                    }
+                    catch(Exception)
```

Comment:
Why?

Reply:
https://github.com/aspnet/IISIntegration/pull/1390#discussion_r217917718

Properties are expected to act like simple public fields. Putting try/catch around every property would be a nuisance.

Reply:
You don't put try/catch around every method call.

Reply:
It also throws in HttpSys: https://github.com/aspnet/HttpSysServer/blob/eba2d7e380fef9f75203bb9a1f4d30827451d512/src/Microsoft.AspNetCore.Server.HttpSys/RequestProcessing/Request.cs#L269

Reply:
You put try/catch around the ones that say they can.
IISIntegration catches client cert errors. https://github.com/aspnet/IISIntegration/blob/5e896ca50678547ce5a71a23edd7dcc5bab45ba7/src/Microsoft.AspNetCore.Server.IISIntegration/ForwardedTlsConnectionFeature.cs#L32-L40 That's a better precident for you here.

Reply:
Will remove try/catch. Creating a new x509 cert will probably not fail as it already passed through IIS cert validation.